Bevy Feathers: an opinionated widget toolkit for building Bevy tooling

viridia opened the pull request:
# Objective

This PR introduces Bevy Feathers, an opinionated widget toolkit and theming system intended for use by the Bevy Editor, World Inspector, and other tools.

The `bevy_feathers` crate is incomplete and hidden behind an experimental feature flag. The API is going to change significantly before release.

viridia committed:
Initial check-in for Feathers.

viridia committed:
Use of Propagate to implement font style inheritance.

viridia committed:
Feathers buttons now calculates background color.

viridia committed:
Slider text now updates.

viridia committed:
Themed font colors, bundle overrides.

viridia committed:
Moved feathers under an experimental feature flag.

viridia committed:
Style tweaks.

viridia committed:
Updated examples, typo, release note.

viridia committed:
Simplified example, fixed warnings.

viridia committed:
Rebuild cargo features.

alice-i-cecile commented on crates/bevy_feathers/src/colors.rs:
```diff
@@ -0,0 +1,28 @@
+pub const LIGHT_GRAY_2: Color = Color::oklcha(0.6106, 0.003, 286.31, 1.0);
+/// White
+pub const WHITE: Color = Color::oklcha(1.0, 0.000000059604645, 90.0, 1.0);
+/// Accent
```
Describing the basic color here, like you did for XYZ, would be helpful.

viridia replied:
I've removed the descriptions in favor of swatches.

BTW, I didn't get the names from the basic colors. I used the color picker app on my phone, but that's not an exact process (depending on illumination).

alice-i-cecile commented on crates/bevy_feathers/src/controls/button.rs:
```diff
@@ -0,0 +1,205 @@
+    pub children: C,
+}
+
+/// Q: How to pass in children?
```
Normal comments, not doc comments for these.

alice-i-cecile commented on crates/bevy_feathers/src/controls/slider.rs:
```diff
@@ -0,0 +1,197 @@
+    /// On-change handler
+    pub on_change: Option<SystemId<In<f32>>>,
+    // Components that are merged into the bundle.
+    // pub overrides: Box<dyn Bundle>,
```
No dead code please.

alice-i-cecile commented on crates/bevy_feathers/src/controls/slider.rs:
```diff
@@ -0,0 +1,197 @@
+    },
+};
+
+/// Slider template properties.
```
I'd appreciate a link to the method where these are used.

alice-i-cecile commented on crates/bevy_feathers/src/controls/slider.rs:
```diff
@@ -0,0 +1,197 @@
+    }
+}
+
+fn update_slider_colors_remove() {}
```
I don't understand why this exists.

viridia replied:
Because I forgot to write this function :)

viridia commented on crates/bevy_feathers/src/colors.rs:
```diff
@@ -0,0 +1,28 @@
+use bevy_color::Color;
```
Another option is for this color set to live alongside the other palettes in bevy_color.

viridia replied:
Done.

IceSentry commented on examples/ui/feathers.rs:
```diff
@@ -0,0 +1,160 @@
+
+fn setup(mut commands: Commands) {
+    // System to print a value when the button is clicked.
+    let on_click = commands.register_system(|| {
```
I think it would be a slightly nicer showcase if the buttons each printed something custom based on what they are. It would also show a bit more how to have multiple different handlers which seems like a more common case.

viridia replied:
Done.

hukasu commented on crates/bevy_feathers/src/handle_or_path.rs:
```diff
@@ -0,0 +1,60 @@
+///
+/// This is useful for when you want to specify an asset, but don't always have convenient
+/// access to an asset server reference.
+#[derive(Clone, Debug)]
+pub enum HandleOrPath<T: Asset> {
+    /// Specify the asset reference as a handle.
+    Handle(Handle<T>),
+    /// Specify the asset reference as a [`String`].
+    Path(String),
+}
```
i wonder if this should be in `bevy_asset`

IceSentry replied:
We do something similar with Shaders. It would be nice for this pattern to be more generalized, but it's a bit out of scope here.

hukasu replied:
i had a suspicion that there was somewhere that already had this, i had in my head that it was in `bevy_gltf` but the type that appears there has more information

alice-i-cecile replied:
Yep, spin this out in follow-up.

hukasu commented on crates/bevy_feathers/src/controls/button.rs:
```diff
@@ -0,0 +1,205 @@
+        props.variant,
+        Hovered::default(),
+        CursorIcon::System(bevy_window::SystemCursorIcon::Pointer),
+        TabIndex(0),
```
Does it make sense to have all buttons in a group with the same index?

viridia replied:
It's fine. Just like on the web, elements with equal tab index are navigated in hierarchy order.

viridia commented on crates/bevy_feathers/src/controls/button.rs:
```diff
@@ -0,0 +1,205 @@
+    /// Rounded corners options
+    pub corners: RoundedCorners,
+    /// Click handler
+    pub on_click: Option<SystemId>,
```
Well, the main reason is that during prototyping you may want to create your layout and don't have your handlers defined yet. I admit that this is a somewhat trivial use case.

Eventually I suspect that `Option<SystemId>` will get replaced with an enum allowing a system id, a cached system, "Ignore", or other options.

alice-i-cecile replied:
Yeah, the `SystemId` based design is the thing I'm most nervous about in this whole design. I'd like to revisit it, but not in this PR.

viridia commented on crates/bevy_feathers/src/colors.rs:
```diff
@@ -0,0 +1,28 @@
+
+/// Black
+pub const BLACK: Color = Color::oklcha(0.0, 0.0, 0.0, 1.0);
+/// Gray 0
```
I want to do this, but here's the issue: because the colors are in oklcha space, we need to do the conversion from that to RGB. Worse, our oklcha color type uses a hue range of 0..1, not 0%..100% as in CSS.

We could use the CSS `oklcha()` function in the HTML snippet, however VSCode is not smart enough to show the preview of the swatch so we'd have to actually generate the docs to see if we got it right.

I don't trust Copilot to do the conversion without error - LLMs are sometimes bad at math.

The right move is probably to write a Python script to do this.

viridia commented on crates/bevy_feathers/src/colors.rs:
```diff
@@ -0,0 +1,28 @@
+use bevy_color::Color;
+
+/// Black
```
I can put in what Rin wrote, although that won't be comprehensive.

viridia replied:
Done.

viridia committed:
Added descriptions to feathers palette, and renamed to "palette".
Simplified slider and button props, default now works.
Removed rough draft comments.
Finished slider styling observer I forgot to implement.
Improved release notes.

viridia committed:
Doc links for template props.

hukasu commented on crates/bevy_feathers/src/palette.rs:
```diff
@@ -0,0 +1,29 @@
+//! The Feathers standard color palette.
+use bevy_color::Color;
+
+/// Black
+pub const BLACK: Color = Color::oklcha(0.0, 0.0, 0.0, 1.0);
+/// Gray 0 - window background
+pub const GRAY_0: Color = Color::oklcha(0.2414, 0.0095, 285.67, 1.0);
+/// Gray 1 - pane background
+pub const GRAY_1: Color = Color::oklcha(0.2866, 0.0072, 285.93, 1.0);
+/// Gray 2 - item background
+pub const GRAY_2: Color = Color::oklcha(0.3373, 0.0071, 274.77, 1.0);
+/// Gray 3 - item background (active)
+pub const GRAY_3: Color = Color::oklcha(0.3992, 0.0101, 278.38, 1.0);
+/// Warm Gray 3 - border
+pub const WARM_GRAY_1: Color = Color::oklcha(0.3757, 0.0017, 286.32, 1.0);
+/// Light Gray 1 - bright label text
+pub const LIGHT_GRAY_1: Color = Color::oklcha(0.7607, 0.0014, 286.37, 1.0);
+/// Light Gray 2 - dim label text
+pub const LIGHT_GRAY_2: Color = Color::oklcha(0.6106, 0.003, 286.31, 1.0);
+/// White - button label text
+pub const WHITE: Color = Color::oklcha(1.0, 0.000000059604645, 90.0, 1.0);
+/// Accent - call-to-action and selection color
+pub const ACCENT: Color = Color::oklcha(0.542, 0.1594, 255.4, 1.0);
+/// Dark Coral - for X-axis inputs and drag handles
+pub const X_AXIS: Color = Color::oklcha(0.5232, 0.1404, 13.84, 1.0);
+/// Olive - for Y-axis inputs and drag handles
+pub const Y_AXIS: Color = Color::oklcha(0.5866, 0.1543, 129.84, 1.0);
+/// Steel Blue - for Z-axis inputs and drag handles
+pub const Z_AXIS: Color = Color::oklcha(0.4847, 0.1249, 253.08, 1.0);
```
These colors are, most of them, related to the colors used by the dark theme, right? other themes would not use most of these, right?

viridia replied:
That decision is above my pay grade...or maybe below it. I don't know the answer :)

Right now I am making as few assumptions as I can.

alice-i-cecile replied:
We're going to focus on a dark theme only for now.

janhohenheim replied:
Could you add the hex code in the comments so that there's a nice preview in VS Code? :)

hukasu commented on crates/bevy_feathers/src/theme.rs:
```diff
@@ -0,0 +1,308 @@
+use bevy_ui::{BackgroundColor, BorderColor};
+
+/// A collection of properties that make up a theme.
+#[derive(Default, Clone)]
+pub struct ThemeProps {
+    /// Map of design tokens to colors.
+    pub color: HashMap<String, Color>,
+    // Other style property types to be added later.
+}
```
Maybe this should be an asset, and have an asset loader, and have the dark theme already as an asset so that people creating custom themes already have an starting point

viridia replied:
It's designed so that it could potentially be an asset - just slap some serde derives on there and you are good to go. But I don't want to do that now. And particularly, for small things like the world inspector, any kind of filesystem dependencies should be avoided. So being able to define themes in code is also a feature.

hukasu commented on crates/bevy_feathers/src/theme.rs:
```diff
@@ -0,0 +1,308 @@
+    pub const BUTTON_PRIMARY_BG_HOVER: &str = "button.primary.bg.hover";
+    /// Primary button background (disabled)
+    pub const BUTTON_PRIMARY_BG_DISABLED: &str = "button.primary.bg.disabled";
+    /// Primary button background (pressed)
+    pub const BUTTON_PRIMARY_BG_PRESSED: &str = "button.primary.bg.pressed";
+    /// Primary button text
+    pub const BUTTON_PRIMARY_TEXT: &str = "button.primary.txt";
+    /// Primary button text (disabled)
+    pub const BUTTON_PRIMARY_TEXT_DISABLED: &str = "button.primary.txt.disabled";
+
+    // Slider
+
+    /// Background for slider
+    pub const SLIDER_BG: &str = "slider.bg";
+    /// Background for slider moving bar
+    pub const SLIDER_BAR: &str = "slider.bar";
+    /// Background for slider moving bar (disabled)
+    pub const SLIDER_BAR_DISABLED: &str = "slider.bar.disabled";
+    /// Background for slider text
+    pub const SLIDER_TEXT: &str = "slider.text";
+    /// Background for slider text (disabled)
+    pub const SLIDER_TEXT_DISABLED: &str = "slider.text.disabled";
+
+    // Checkbox
+
+    /// Checkbox border around the checkmark
+    pub const CHECKBOX_BORDER: &str = "checkbox.border";
+    /// Checkbox border around the checkmark (hovered)
+    pub const CHECKBOX_BORDER_HOVER: &str = "checkbox.border.hover";
+    /// Checkbox border around the checkmark (disabled)
+    pub const CHECKBOX_BORDER_DISABLED: &str = "checkbox.border.disabled";
+    /// Checkbox check mark
+    pub const CHECKBOX_MARK: &str = "checkbox.mark";
+    /// Checkbox check mark (disabled)
+    pub const CHECKBOX_MARK_DISABLED: &str = "checkbox.mark.disabled";
+    /// Checkbox label text
+    pub const CHECKBOX_TEXT: &str = "checkbox.text";
+    /// Checkbox label text (disabled)
+    pub const CHECKBOX_TEXT_DISABLED: &str = "checkbox.text.disabled";
+}
```
are these generic enough that any app would use them or are these more related to the editor?

viridia committed:
Fix CI problems.

IceSentry commented on examples/ui/feathers.rs:
```diff
@@ -0,0 +1,180 @@
+                    children![
+                        button(
+                            ButtonProps {
+                                on_click: Some(commands.register_system(|| {
```
I haven't kept up with all the various ecs and bsn work. So out of curisoity, is there a future where we'll be able to do something like this:

```rust
ButtonProps {
	click_handler: || { 
		info!("clicked!");
	},
	..default()
}
```

alice-i-cecile replied:
Yes, I would like to be able to do that. I need to take a proper crack at the code here though.

ickshonpe commented on crates/bevy_feathers/src/font_styles.rs:
```diff
@@ -0,0 +1,62 @@
+}
+
+/// An observer which looks for changes to the `InheritableFont` component on an entity, and
+/// propagates downward the text color to all participating text entities.
```
This propagates the font, not the text color?

ickshonpe commented on crates/bevy_feathers/src/controls/slider.rs:
```diff
@@ -0,0 +1,214 @@
+        BackgroundGradient(vec![Gradient::Linear(LinearGradient {
+            angle: PI * 0.5,
+            stops: vec![
+                ColorStop::new(Color::NONE, Val::Percent(0.)),
+                ColorStop::new(Color::NONE, Val::Percent(50.)),
+                ColorStop::new(Color::NONE, Val::Percent(50.)),
+                ColorStop::new(Color::NONE, Val::Percent(100.)),
```
The the zero and one hundred percent stops can be elided here. When there is space before the first stop or after the last stop it fills with the respective stop's color.

alice-i-cecile replied:
I like that this is more explicit.

ickshonpe replied:
I guess so, but it's also a little more complicated to update the colors as there are four values to change instead of two.

viridia committed:
Added color swatches.

viridia committed:
Fixed lying comment.

IceSentry reviewed (approved):
I don't know how I feel about the embedded fonts, but since it's behind a feature it doesn't matter that much.

Other than that LGTM

alice-i-cecile commented on crates/bevy_feathers/src/controls/button.rs:
```diff
@@ -0,0 +1,205 @@
+
+/// Parameters for the button template, passed to [`button`] function.
+#[derive(Default, Clone)]
+pub struct ButtonProps {
```
I'm surprised that text labels aren't part of these props / spawned with the button. What are your plans there?

viridia replied:
Labels are just children. The reason for not making the label part of ButtonProps is that a button might not have a label, it might have an icon. Or a label and an icon. Or a drop-down arrow. Or a badge indicating "unread notifications". There are too many possibilities to systematize.

alice-i-cecile commented on crates/bevy_feathers/src/controls/mod.rs:
```diff
@@ -0,0 +1,16 @@
+pub use button::{button, ButtonPlugin, ButtonProps, ButtonVariant};
+pub use slider::{slider, SliderPlugin, SliderProps};
+
+/// Plugin which registers all feathers controls.
```
```suggestion
/// Plugin which registers all `bevy_feathers` controls.
```

alice-i-cecile commented on crates/bevy_feathers/src/controls/slider.rs:
```diff
@@ -0,0 +1,214 @@
+};
+
+/// Slider template properties, passed to [`slider`] function.
+// TODO: These should all be replaced with reactive signals.
```
Not convinced that this TODO is net helpful: that work is fairly blocked, so it's mostly just going to be noise.

alice-i-cecile commented on crates/bevy_feathers/src/cursor.rs:
```diff
@@ -0,0 +1,69 @@
+pub struct DefaultCursorIcon(pub CursorIcon);
+
+/// System which updates the window cursor icon whenever the mouse hovers over an entity with
+/// a [`CursorIcon`] component. If no entity is hovered, the cursor icon is set to
```
Follow-up: we should be updating the docs for `bevy::window::CursorIcon` to signal to users that this is an intended pattern, and moving this code out into `bevy_picking` or something.

alice-i-cecile commented on crates/bevy_feathers/src/font_styles.rs:
```diff
@@ -0,0 +1,62 @@
+use crate::handle_or_path::HandleOrPath;
+
+/// A component which, when inserted on an entity, will load the given font and propagate it
+/// downward to any child text entity that has the `UseTheme` marker.
```
```suggestion
/// downward to any child text entity that has the [`UseTheme`] marker.
```

alice-i-cecile commented on crates/bevy_feathers/src/theme.rs:
```diff
@@ -0,0 +1,308 @@
+#[component(immutable)]
+pub struct ThemeFontColor(pub &'static str);
+
+/// A marker component that is used to indicate that the text entity wants to opt-in to using
```
The naming feels at odds with the very narrow scope. I'd either rename this or tweak the docs to cover the intended scope, then say "currently this only does inherited text styling".

alice-i-cecile commented on crates/bevy_feathers/src/theme.rs:
```diff
@@ -0,0 +1,308 @@
+/// A collection of properties that make up a theme.
+#[derive(Default, Clone)]
+pub struct ThemeProps {
+    /// Map of design tokens to colors.
```
Linking to module docs for design tokens would be helpful here.

alice-i-cecile commented on crates/bevy_feathers/src/palette.rs:
```diff
@@ -0,0 +1,29 @@
+//! The Feathers standard color palette.
```
I would appreciate an explanation of how these relate to design tokens here.

alice-i-cecile commented on crates/bevy_feathers/Cargo.toml:
```diff
@@ -0,0 +1,40 @@
+name = "bevy_feathers"
+version = "0.16.0-dev"
+edition = "2024"
+description = "An collection of UI widgets for building editors and utilities in Bevy"
```
```suggestion
description = "A collection of UI widgets for building editors and utilities in Bevy"
```

alice-i-cecile commented on crates/bevy_feathers/src/theme.rs:
```diff
@@ -0,0 +1,308 @@
+#[derive(Component)]
+pub struct UseTheme;
+
+/// UX design tokens
```
This needs a proper definition / justification here. We have a lot of open source backend folks and game devs, who will have never seen this concept before.

alice-i-cecile commented on release-content/release-notes/feathers.md:
```diff
@@ -0,0 +1,24 @@
+pull_requests: [19730]
+---
+
+As part of our commitment to enhance the developer experience for UI creation in Bevy, we are
+pleased to introduce "Feathers" - a comprehensive widget set that offers:
```
```suggestion
To make it easier for Bevy engine developers and third-party tool creators to make comfortable, visually cohesive tooling,
we're pleased to introduce "Feathers" - a comprehensive widget set that offers:
```

alice-i-cecile commented on release-content/release-notes/feathers.md:
```diff
@@ -0,0 +1,24 @@
+
+Feathers isn't meant as a toolkit for building exciting and cool game UIs: it has a somewhat plain
+and utilitarian look and feel suitable for editors and graphical utilities. That being said, using
+the themeing framework, you can spice up the colors quite a bit.
```
```suggestion
the themeing framework, you can spice up the colors quite a bit.
It can also serve as a helpful base to understand how to extend and style `bevy_ui` and our new core widgets;
copy the code into your project and start hacking!
```

alice-i-cecile commented on crates/bevy_feathers/src/lib.rs:
```diff
@@ -0,0 +1,66 @@
+//! `bevy_feathers` is a collection of styled and themed widgets for building editors and
+//! inspectors.
```
```suggestion
//! `bevy_feathers` is a collection of styled and themed widgets for building editors and
//! inspectors.
//! 
//! The aesthetic choices made here are designed with a future Bevy Editor in mind,
//! but this crate is deliberately exposed to the public to allow the broader ecosystem to easily create
//! tooling for themselves and others that fits cohesively together.
//!
//! While it may be tempting to use this crate for your game's UI, it's deliberately not intended for that.
//! We've opted for a clean, functional style, and prioritized consistency over customization.
//! That said, if you like what you see, it can be a helpful learning tool.
//! Consider copying this code into your own project,
//! and refining the styles and abstractions provided to meet your needs.
//! 
//! ## Warning: Experimental!
//! All that said, this crate is still experimental and unfinished!
//! It will change in breaking ways, and there will be both bugs and limitations.
//! 
//! Please report issues, submit fixes and propose changes.
//! Thanks for stress-testing; let's build something better together.
```

alice-i-cecile reviewed (changes requested):
I've left some suggestions / requests, mostly for better docs, but I'm happy with the general shape of this and the initial style choices made.

I want to get the module structure / code conventions a bit cleaner before we merge this to avoid maintenance pain in the future, but once that's done, I think we should merge this and start iterating on the functionality and aesthetics in smaller PRs.

I've also written some proper module docs for you here because I could <3

viridia committed:
Update doc comments from code review.

Co-authored-by: Alice Cecile <alice.i.cecile@gmail.com>

viridia committed:
From review feedback:
* Break up theme module
* Improved module docs
* Rename "dark" to "dark_theme"

alice-i-cecile reviewed (approved):
Good :) I'm happy with the organization and docs now. I know @cart wants to take a look first, so I'll mark this as S-Needs-SME and push for a decision :)

A couple of non-blocking suggestions that you can take or leave:
- I would probably rename `theme.rs` -> `theming.rs`, to make clear that that module has a "theming framework" in it.
- I would group `tokens`/dark_theme`/`palette`/`constants` into a `aesthetics` or `design` module (but keep them as separate files): they're all aesthetic tuning levers, so it feels like they belong together.

hukasu commented:
`S-Needs-🛒`

cart commented on crates/bevy_feathers/src/assets/fonts/FiraSans-License.txt:
```diff
@@ -0,0 +1,93 @@
+Copyright (c) 2012-2015, The Mozilla Foundation and Telefonica S.A.
```
I think its worth considering [Firava](https://github.com/hellogreg/firava) (a variable font variant of Fira Sans), rather than checking in each font variant. Provided it works, it would give users maximum flexibility and (ideally) cut down on a bit on complexity.
I believe we can render variable fonts, however it looks like we aren't currently wiring up "font width", so idk how functional it would be in the short term. We don't _need_ to resolve this now. Although committing to that path now would mean we aren't adding these fonts to the history, which would be nice.

cart commented on crates/bevy_feathers/src/font_styles.rs:
```diff
@@ -0,0 +1,62 @@
+
+use crate::handle_or_path::HandleOrPath;
+
+/// A component which, when inserted on an entity, will load the given font and propagate it
```
This makes sense given the current behavior (both lack of font inheritance and wanting to express an AssetPath instead of needing to manually load a handle on an asset server, which BSN will fix). I think we might want to consider making "text style" inheritance happen implicitly by default, but thats not something to resolve here.

viridia replied:
Sure, let's leave this for a follow-up.

cart reviewed (commented):
Love how this looks and behaves! I think this is largely ready to go. My biggest hesitance is the inclusion of the theme framework.

viridia committed:
Fix broken example.

viridia commented on crates/bevy_feathers/src/handle_or_path.rs:
```diff
@@ -0,0 +1,61 @@
+    /// Specify the asset reference as a handle.
+    Handle(Handle<T>),
+    /// Specify the asset reference as a [`String`].
+    Path(String),
```
I'm fine with changing it.

viridia commented on crates/bevy_feathers/src/theme.rs:
```diff
@@ -0,0 +1,114 @@
+use bevy_ui::{BackgroundColor, BorderColor};
+
+/// A collection of properties that make up a theme.
+#[derive(Default, Clone)]
```
For the Bevy editor, what we want eventually (according to Alice's comment) is a light / dark mode switch. I don't see the need for arbitrary color schemes a la VSCode. I don't plan to implement a light mode any time soon, but I think there's value in having the infrastructure in place now instead of adding it later as an afterthought.

Also, a BSN-based theming solution might not be all that different. While it might not use "theming components" (like `ThemedBackgroundColor` and such), the basic idea of a table of properties in a resource seems uncontroversial. That is, unless you are planning on supporting contextual overrides, which is a much harder problem - I did an informal survey in editor-dev of what people wanted/needed, and it sounded like there wasn't much interest in contextual themes. A single global color scheme which can be changed at runtime is all that folks seem to be asking for.

viridia replied:
@cart One problem with defining color constants is that we can't mix colors in constant initalizers. If, for example, that if we want GRAY_1 but with an alpha of 0.5, we have to make a new color definition for that, and it's no longer connected to the original GRAY_1.

viridia committed:
Added feathers namespace to design tokens.

viridia committed:
Fix doc link.

viridia committed:
Fix doc link.

viridia committed:
Add additional contributor

alice-i-cecile commented:
I'm happy with this as a base. There's lots to explore and follow-up on, but let's get this merged and let people start experimenting :)
